File: typescript/submarine.ts
function getInput() {
  return `
        forward 5
        down 5
        forward 8
        up 3
        down 8
        forward 2
    `;
}

function parseLine(line: string): [number, number] {
  const [direction, distance] = line.trim().split(' ');

  if (!direction || !direction) return [0, 0];

  const amount = parseInt(distance);

  if (direction === 'forward') {
    return [amount, 0];
  } else if (direction === 'up') {
    return [0, -amount];
  }
  return [0, amount];
}

export function Run() {
  const result = getInput()
    .split('\n')
    .map((line) => parseLine(line))
    .reduce(
      (acc, curr) => {
        acc[0] += curr[0];
        acc[1] += curr[1];
        return acc;
      },
      [0, 0]
    );

  console.log(result, result[0] * result[1]);
}

File: rust/submarine.rs
pub fn get_input() -> &'static str {
    return "
		    forward 5
        down 5
        forward 8
        up 3
        down 8
        forward 2
	";
}

#[derive(Debug)]
struct Point {
    x: i32,
    y: i32,
}

fn parse_line(line: &str) -> Point {
    let (direction, amount) = str::trim(line).split_once(" ").expect("Must contain a whitespace");

    if direction.is_empty() || amount.is_empty() {
        return Point { x: 0, y: 0 };
    }

    let amount = str::parse::<i32>(amount).expect("Second arg must be an integer");

    if direction == "forward" {
        return Point { x: amount, y: 0 };
    } else if direction == "up" {
        return Point { x: 0, y: -amount };
    }

    return Point { x: 0, y: amount };
}

pub fn run() {
    let result =
        str::trim(get_input())
            .lines()
            .map(parse_line)
            .fold(Point { x: 0, y: 0 }, |mut acc, point| {
                acc.x += point.x;
                acc.y += point.y;

                return acc;
            });

    println!("{:?} {1}", result, result.x * result.y)
}
